File: Balance a Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:
        arr = []

        def read(node):
            nonlocal arr
            if node is None:
                return

            read(node.left)
            arr.append(node.val)
            read(node.right)

        read(root)

        def balance(left, right):
            nonlocal arr
            if left == right:
                return TreeNode(val=arr[left])

            if left > right:
                return None

            mid = (left + right) // 2
            node = TreeNode(val=arr[mid])
            node.left = balance(left, mid - 1)
            node.right = balance(mid + 1, right)
            return node

        return balance(0, len(arr) - 1)

File: Balance a Binary Search Tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
func read(node* TreeNode, arr *[]int) {
    if node != nil {
        read(node.Left, arr)
        *arr = append(*arr, node.Val)
        read(node.Right, arr)
    }
}

func balance(left int, right int, arr *[]int) *TreeNode {
    if left == right {
        return &TreeNode{Val: (*arr)[left]}
    }

    if left > right {
        return nil
    }

    mid := (left+right)/2
    node := TreeNode{Val: (*arr)[mid]}
    node.Left = balance(left, mid-1, arr)
    node.Right = balance(mid+1, right, arr)
    return &node
}

func balanceBST(root *TreeNode) *TreeNode {
    arr := make([]int, 0)
    read(root, &arr)
    return balance(0, len(arr)-1, &arr)
}
